### PR TITLE
Region auto enable on env variable

### LINKED ISSUE #7328: [Feature Request] Region auto enable on env variable
## Region auto-enable

1. On creation of a ConfidentialClientApplication, MSAL shall detect env variable `MSAL_FORCE_REGION`, which will be set to a specific region (e.g. `westus1`)
2. If this env variable is set, MSAL shall opt-in to ESTS-R with the value of this variable.

## Disabling the behavior

If region is explicitly set in MSAL to the string "DisableMsalForceRegion`, then the above no longer applies

## Conflicts with `WithAzureRegion(xyz)`

Use of the api `WithRegion(xyz)` takes precedence over the env variable. 

## Acceptance tests

For all of the tests, assume env variable MSAL_FORCE_REGION=eastus

1. No other config is used. ACTUAL region used: eastus
2. App developer configures region "westus" in MSAL. ACTUAL region used: westus
3. App developer configures region "DisableMsalForceRegion" in MSAL. ACTUAL region used: none

 ### Original issue: 
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/4930

### PR DESCRIPTION
**Pull request by Robbie-Microsoft**
On creation of a ConfidentialClientApplication, MSAL shall detect env variable MSAL_FORCE_REGION, which will be set to a specific region (e.g. westus1)

If this env variable is set, MSAL shall opt-in to ESTS-R with the value of this variable.

This feature request is located [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/7328).

### CODE REVIEW

File: lib/msal-node/test/client/ConfidentialClientApplication.spec.ts [hunk 1]
```diff
@@ -340,6 +343,89 @@ describe("ConfidentialClientApplication", () => {
...
+                expect(acquireTokenByClientCredentialSpy).toHaveBeenCalledTimes(
+                    1
+                );
+                expect(
+                    buildOauthClientConfigurationSpy.mock.lastCall[3]
+                        .azureRegion
+                ).toEqual(region);
+            });
+
+            test('region is not passed in through the request, the MSAL_FORCE_REGION environment variable is set to "DisableMsalForceRegion"', async () => {
```
Comment (bgavrilMS) [inline lib/msal-node/test/client/ConfidentialClientApplication.spec.ts:438]: First 2 tests are spot on, but for the 3rd one there is confusion about how app dev configures this magic string `DisableMsalForceRegion`

The spec states: 

image

In other words, if the env variable `MSAL_FORCE_REGION=eastus` and app developer sets the region programatically (like at line 396) to `DisableMsalForceRegion` then you are to ignore ESTS-R and go to global.

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
See comment about disable

File: lib/msal-node/test/client/ConfidentialClientApplication.spec.ts (hunk 1 above)
Comment (Robbie-Microsoft) [reply lib/msal-node/test/client/ConfidentialClientApplication.spec.ts:438]: Based on your third acceptance test as listed in [the feature request](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/7328): if the dev sets the region to "DisableMsalForceRegion", then even when the env variable is defined, the region should be set to `undefined`. Is that correct?

Comment (bgavrilMS) [reply lib/msal-node/test/client/ConfidentialClientApplication.spec.ts:438]: Correct. See .NET PR: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4954

File: lib/msal-node/src/client/ConfidentialClientApplication.ts [hunk 2]
```diff
@@ -136,8 +138,25 @@ export class ConfidentialClientApplication
             );
         }
 
+        /*
+         * if this env variable is set, and the developer provided region isn't defined and isn't "DisableMsalForceRegion",
```
Comment (neha-bhargava) [inline lib/msal-node/src/client/ConfidentialClientApplication.ts:142]:
```suggestion
         * If this env variable is set, and the developer provided region isn't defined and isn't "DisableMsalForceRegion",
```

File: lib/msal-node/src/client/ConfidentialClientApplication.ts [hunk 3]
```diff
@@ -136,8 +138,25 @@ export class ConfidentialClientApplication
...
+            process.env[MSAL_FORCE_REGION];
+
+        let region: AzureRegion | undefined;
+        if (validRequest.azureRegion !== "DisableMsalForceRegion") {
+            region =
+                !validRequest.azureRegion && ENV_MSAL_FORCE_REGION
+                    ? ENV_MSAL_FORCE_REGION
+                    : validRequest.azureRegion;
+        } else {
+            validRequest.azureRegion = undefined;
```
Comment (rayluo) [inline lib/msal-node/src/client/ConfidentialClientApplication.ts]: Although the test cases are valid and currently passing, the implementation pattern here still looks weird:

```typescript
let sentinel | undefined;

if (condition) {sentinel = something;} else {variable_BAR = another_thing;}

use(sentinel);
```

I suppose this branch could/should be adjusted like this?

```suggestion
            region = undefined;
```

Comment (Robbie-Microsoft) [reply lib/msal-node/src/client/ConfidentialClientApplication.ts]: oops. good catch, thanks

**Review by rayluo — APPROVED** [review]

File: lib/msal-node/src/client/ConfidentialClientApplication.ts [hunk 4]
```diff
@@ -136,8 +138,23 @@ export class ConfidentialClientApplication
...
+        /*
+         * if this env variable is set, and the developer provided region isn't defined and isn't "DisableMsalForceRegion",
+         * MSAL shall opt-in to ESTS-R with the value of this variable
+         */
+        const ENV_MSAL_FORCE_REGION: AzureRegion | undefined =
+            process.env[MSAL_FORCE_REGION];
+
+        let region: AzureRegion | undefined;
+        if (validRequest.azureRegion !== "DisableMsalForceRegion") {
+            region =
```
Comment (bgavrilMS) [inline lib/msal-node/src/client/ConfidentialClientApplication.ts]: nit: refactor to use "if"... this is quite difficult to read.

File: lib/msal-node/test/client/ConfidentialClientApplication.spec.ts [hunk 5]
```diff
@@ -340,6 +343,87 @@ describe("ConfidentialClientApplication", () => {
             );
         });
 
+        describe("region is determined correctly", () => {
+            let acquireTokenByClientCredentialSpy: jest.SpyInstance;
+            let buildOauthClientConfigurationSpy: jest.SpyInstance;
+            let client: ConfidentialClientApplication;
+            let request: ClientCredentialRequest;
+            beforeEach(() => {
+                acquireTokenByClientCredentialSpy = jest.spyOn(
```
Comment (bgavrilMS) [inline lib/msal-node/test/client/ConfidentialClientApplication.spec.ts:368]: Why not use HTTP mocking and assert that the token request goes over the ESTS-R token endpoint?

**Review by bgavrilMS — APPROVED** [review]

**Review by bgavrilMS — COMMENTED** [review]
tests look great now, thanks!

**Pull request merged**